### PR TITLE
Text report -> Html report

### PR DESCRIPTION
**Pull request by cvaldez98**
This makes reporting better by making the switch over to an HTML report page as opposed to a plain text one. 
- The actual HTML templating can be iterated on and currently I didn't really spend much time on making the webpage look _nice_, I mostly just worked on getting the json2html transform to work correctly and to promptly remove the old code and replace it.

### CODE REVIEW

File: src/report.ts [hunk 1]
```diff
@@ -1,60 +1,68 @@
...
-
-      text += `\n`;
-    }
-  }
-  appState.report += text;
-  return true;
+  input.forEach( (currSuiteResult) => {
+    // append to the end of the list. 
+    appState.reportJSON.push(currSuiteResult)
+  })
```
Comment (felixrieseberg) [inline src/report.ts]: I think `prettier` didn't run?

File: src/report.ts [hunk 2]
```diff
@@ -1,60 +1,68 @@
...
+  const template = fs.readFileSync(path.resolve(dirname, './report/static/template.html'), 'utf8');
+  const html = convert(report);
+  const result = template.replace('HTMLHERE', html);
   const reportPath = path.join(appState.absPath, appState.fileName);
-  return fs.writeFile(reportPath, appState.report);
+  fs.writeFileSync(reportPath, result);
+}
+
+function convert(report:string) {
+  const json2html = require('node-json2html');
```
Comment (felixrieseberg) [inline src/report.ts]: Could you return at the top with an `import`?

File: src/report.ts [hunk 3]
```diff
@@ -1,60 +1,68 @@
...
+  const result:Result = {'ok': succeeded, 'name': input.name, 'error': error, 'skipped': false};
+  const suiteResult:SuiteResult = {'name': text, 'results':[result]};
+  appState.reportJSON.push(suiteResult);
 }
 
+// Write our JSON data to HTML
 export function writeToFile() {
+  const JSONreport = JSON.stringify(appState.reportJSON);
+  console.log('JSONreport', JSONreport);
+  createPage(JSONreport);
```
Comment (felixrieseberg) [inline src/report.ts]: I'd suggest catching write errors - either here or where you actually call the `fs.writeFile`.

File: src/report.ts [hunk 4]
```diff
@@ -1,60 +1,84 @@
...
+  }
 
-  appState.report += text;
+  return true;
 }
 
-export function writeToFile() {
+// Create HTML report page
+function createPage(report: string) {
+  const dirname = '/Users/cvaldez/Documents/slack-old-joe/src/';
```
Comment (felixrieseberg) [inline src/report.ts]: That doesn't look right

File: src/report.ts [hunk 5]
```diff
@@ -1,60 +1,84 @@
...
   const reportPath = path.join(appState.absPath, appState.fileName);
-  return fs.writeFile(reportPath, appState.report);
+  fs.writeFileSync(reportPath, result);
+}
+
+function convert(report: string) {
+  // Transform our list of tests into HTML.
+  const transforms = {
+    '<>': 'h1',
+    text: '${name}\n',
```
Comment (felixrieseberg) [inline src/report.ts]: Is that unique to how the transforms work or did you mean to use backticks?

Comment (cvaldez98) [reply src/report.ts]: That is unique!

File: src/report.ts (hunk 4 above)
Comment (cvaldez98) [reply src/report.ts]: Thanks for catching this 🙇

File: src/ambient.d.ts [hunk 6]
```diff
@@ -3,3 +3,4 @@ declare module 'applescript';
 declare module 'electron-create-menu';
 declare module 'electron-context-menu';
 declare module 'react-stay-scrolled';
+declare module 'node-json2html'
```
Comment (felixrieseberg) [inline src/ambient.d.ts]: `;` missing here

File: src/report.ts [hunk 7]
```diff
@@ -1,60 +1,83 @@
...
+  succeeded: boolean,
+  error?: string
 ) {
   let text = `\n\n# Slack Old Joe Run ${input.name} (previously failed test)\n`;
-  text += `-`.padEnd(50, '-');
-  text += `\n\n`;
+  const result: Result = {
+    ok: succeeded,
+    name: input.name,
+    error: error,
```
Comment (felixrieseberg) [inline src/report.ts]:
```suggestion
    error,
```

File: src/report.ts [hunk 8]
```diff
@@ -1,60 +1,83 @@
...
+    text: '${name}\n',
+    html: function(): any {
+      return json2html.transform((this as any).results, transforms.child);
+    },
+    child: {
+      '<>': 'li',
+      style: 'font-size:16px',
+      text: '${name} Passed: ',
+      html: function(): any {
+        return (this as any).error
```
Comment (felixrieseberg) [inline src/report.ts]: What's going on with the many `any` declarations here?

Comment (cvaldez98) [reply src/report.ts]: typescript will give out an error without casting `this` as `any`. This is because the javascript module will json2html has no type support and has an implicit `this` since we have a nested transform. 
TL;DR it will still work without the casting but typescript will complain and have an error.

Comment (felixrieseberg) [reply src/report.ts]: There's a [solution to that problem](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters) (which I think we should use), but let's merge this in first and then tackle the `this` here 🙇
